feat(docs): shared object example (#4789)
iotaledger/iota: docs/content/developer/iota-101/objects/shared-object-example.mdx

[`SharedCoins`](https://github.com/iotaledger/iota/blob/0494dad2b1d64c054795bac75342597e2eda1e9a/examples/move/transfer-to-shared-object/sources/shared_coins.move#L11-L14) will hold IOTA coins, enabling anyone to deposit and withdraw coins from it.

### [`create`](https://github.com/iotaledger/iota/blob/0494dad2b1d64c054795bac75342597e2eda1e9a/examples/move/transfer-to-shared-object/sources/shared_coins.move#L16-L22)

This function initializes a shared object with an empty array named `coins`.

```move reference
https://github.com/iotaledger/iota/blob/0494dad2b1d64c054795bac75342597e2eda1e9a/examples/move/transfer-to-shared-object/sources/shared_coins.move#L16-L22
```

### [`delete`](https://github.com/iotaledger/iota/blob/0494dad2b1d64c054795bac75342597e2eda1e9a/examples/move/transfer-to-shared-object/sources/shared_coins.move#L25-L29)

> This function removes the `shared_coins` shared object and destroys all the IOTA coins it contains.

Reviewer: ```suggestion
This function destroys the `SharedCoins` shared object.
```
The coins need to be moved out before, they can't be destroyed

```move reference
https://github.com/iotaledger/iota/blob/0494dad2b1d64c054795bac75342597e2eda1e9a/examples/move/transfer-to-shared-object/sources/shared_coins.move#L25-L29
```

### [`deposit_coin`](https://github.com/iotaledger/iota/blob/0494dad2b1d64c054795bac75342597e2eda1e9a/examples/move/transfer-to-shared-object/sources/shared_coins.move#L31-L38)

This function deposits IOTA coins into `SharedCoins` by adding them to the `coins` array of `SharedCoins`.  

> Before using this function, ensure that some IOTA coins have been transferred to address of the `SharedCoins` object. For the `coin` parameter, provide the address of the IOTA coin you transferred to `SharedCoins`. For further details, refer to [Receiving](../../../references/framework/iota-framework/transfer.mdx#0x2_transfer_Receiving) and [public_receive](../../../references/framework/iota-framework/transfer.mdx#function-public_receive).

Reviewer: Sorry I forgot `the`, also not sure how clear "address" is here, every object ID is an address, but might be not clear for everyone
```suggestion
Before using this function, ensure that some IOTA coins have been transferred to the address/object ID of the `SharedCoins` object.
For the `coin` parameter, provide the address/object ID of the IOTA coin you transferred to `SharedCoins`. For further details, refer to [Receiving](../../../references/framework/iota-framework/transfer.mdx#0x2_transfer_Receiving) and [public_receive](../../../references/framework/iota-framework/transfer.mdx#function-public_receive).
```

> Before using this function, ensure that some IOTA coins have been transferred to `SharedCoins`. For the `coin` parameter, provide the address of the IOTA coin you transferred to `SharedCoins`. For further details, refer to [Receiving](../../../references/framework/iota-framework/transfer.mdx#0x2_transfer_Receiving) and [public_receive](../../../references/framework/iota-framework/transfer.mdx#function-public_receive).

Reviewer: ```suggestion
Before using this function, ensure that some IOTA coins have been transferred to address of the `SharedCoins` object. For the `coin` parameter, provide the address of the IOTA coin you transferred to `SharedCoins`. For further details, refer to [Receiving](../../../references/framework/iota-framework/transfer.mdx#0x2_transfer_Receiving) and [public_receive](../../../references/framework/iota-framework/transfer.mdx#function-public_receive).
```

---
title: Shared Object Example
description: An example of how to efficiently utilize shared objects.
---

# Shared Object Example

> In previous tutorial we have discussed about comparision of shared object with owned object. In this tutorial we will write an contract to understand more about shared objects.

Reviewer 2: In the previous tutorial, we discussed comparing a shared object with an owned object. In this tutorial, we will create a smart contract to gain a deeper understanding of shared objects.

## Create Move Package

To [create a new move package](../../getting-started/create-a-package.mdx) run the following command:

```shell
iota move new transfer-to-shared-object
```
## Package Overview

In this example, we will create a package that allows users to create new shared objects of type [`SharedCoins`](https://github.com/iotaledger/iota/blob/0494dad2b1d64c054795bac75342597e2eda1e9a/examples/move/transfer-to-shared-object/sources/shared_coins.move#L11-L14), delete them, add new coins to the [`SharedCoins`](https://github.com/iotaledger/iota/blob/0494dad2b1d64c054795bac75342597e2eda1e9a/examples/move/transfer-to-shared-object/sources/shared_coins.move#L11-L14), and transfer coins to a recipient address from a [`SharedCoins`](https://github.com/iotaledger/iota/blob/0494dad2b1d64c054795bac75342597e2eda1e9a/examples/move/transfer-to-shared-object/sources/shared_coins.move#L11-L14).  

> [`SharedCoins`](https://github.com/iotaledger/iota/blob/0494dad2b1d64c054795bac75342597e2eda1e9a/examples/move/transfer-to-shared-object/sources/shared_coins.move#L11-L14) will hold IOTA coins, enabling anyone to deposit and withdraw coins from it.

Reviewer 2: Can you add a space between lines 19 and 20? It looks better this way when previewed. You can also just merge both lines into one, but I see that `SharedCoins` will be repeated twice in one line.

Author: sure, thanks for the suggesstion.

---
title: Shared Object Example
description: An example of how to efficiently utilize shared objects.
---

# Shared Object Example

> In the previous tutorial, we discussed comparing a shared object with an owned object. In this tutorial, we will create a smart contract to gain a deeper understanding of shared objects.

Reviewer 3: ```suggestion
In the previous sections, we discussed comparing a shared object with an owned object. In this guide, we will create a smart contract that can receive objects to a shared object and transfer them to gain a deeper understanding of shared objects.
```
Please add a link to the previous section, the user may land here from a search engine.

### [`create`](https://github.com/iotaledger/iota/blob/0494dad2b1d64c054795bac75342597e2eda1e9a/examples/move/transfer-to-shared-object/sources/shared_coins.move#L16-L22)

This function initializes a shared object with an empty array named `coins`.

```move reference
https://github.com/iotaledger/iota/blob/0494dad2b1d64c054795bac75342597e2eda1e9a/examples/move/transfer-to-shared-object/sources/shared_coins.move#L16-L22
```

### [`delete`](https://github.com/iotaledger/iota/blob/0494dad2b1d64c054795bac75342597e2eda1e9a/examples/move/transfer-to-shared-object/sources/shared_coins.move#L25-L29)

This function removes the `shared_coins` shared object and destroys all the IOTA coins it contains.

```move reference
https://github.com/iotaledger/iota/blob/0494dad2b1d64c054795bac75342597e2eda1e9a/examples/move/transfer-to-shared-object/sources/shared_coins.move#L25-L29
```

### [`deposit_coin`](https://github.com/iotaledger/iota/blob/0494dad2b1d64c054795bac75342597e2eda1e9a/examples/move/transfer-to-shared-object/sources/shared_coins.move#L31-L38)

This function deposits IOTA coins into `SharedCoins` by adding them to the `coins` array of `SharedCoins`.  

> Before using this function, ensure that some IOTA coins have been transferred to `SharedCoins`. For the `coin` parameter, provide the address of the IOTA coin you transferred to `SharedCoins`. For further details, refer to [Receiving](https://docs.iota.org/references/framework/iota-framework/transfer#0x2_transfer_Receiving) and [public_receive](https://docs.iota.org/references/framework/iota-framework/transfer#function-public_receive).

Reviewer 3: Please use relative links for internal docs

Author: sure.

```move reference
https://github.com/iotaledger/iota/blob/0494dad2b1d64c054795bac75342597e2eda1e9a/examples/move/transfer-to-shared-object/sources/shared_coins.move#L31-L38
```

### [`transfer_coin`](https://github.com/iotaledger/iota/blob/0494dad2b1d64c054795bac75342597e2eda1e9a/examples/move/transfer-to-shared-object/sources/shared_coins.move#L41-L48)

> This function transfers an IOTA coin to the specified `recipient` address. It removes the coin object from the `coins` array and utilizes the `public_transfer` function to complete the transfer.

Reviewer 3: link to public_transfer

Author: sure, will update in the next commit

```move reference
https://github.com/iotaledger/iota/blob/0494dad2b1d64c054795bac75342597e2eda1e9a/examples/move/transfer-to-shared-object/sources/shared_coins.move#L41-L48
```

> - You can view the complete contract code [here](https://github.com/iotaledger/iota/blob/dev-tools/transfer-to-shared-object-example/examples/move/transfer-to-shared-object/sources/shared_coins.move).

Reviewer 3: dead link

Author: resolved

## Conclusion

> In this example, you explored the implementation of shared objects and gained an understanding of how shared objects function on-chain.

Reviewer 3: ```suggestion
```
I'm not sure we need this.
